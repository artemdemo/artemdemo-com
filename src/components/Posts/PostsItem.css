/*
    See explanation in the code PostsItem.jsx
*/
code {
    padding: .1em;
    border-radius: .3em;
    color: black;
    background: #f5f2f0;
    text-shadow: 0 1px white;
    font-family: Consolas, Monaco, 'Andale Mono', 'Ubuntu Mono', monospace;
}