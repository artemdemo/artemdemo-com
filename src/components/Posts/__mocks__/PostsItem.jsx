import React from 'react';

const PostsItem = props => (
    <div data-mock='PostsItem'>
        {JSON.stringify(props)}
    </div>
);

export default PostsItem;