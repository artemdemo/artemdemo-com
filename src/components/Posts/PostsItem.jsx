import React from 'react';
import PropTypes from 'prop-types';
import {Link} from 'gatsby';
import _get from 'lodash/get';
import PostDate from '../PostDate/PostDate';
import * as utils from '../../services/utils';
import {renderMd} from '../../services/md';

// Yes, I know, not the best solution.
// The problem is that I need to use here custom service for markdown,
// because `gatsby-transformer-remark` will work only for posts, but _not_ for excerpts.
import './PostsItem.css';

class PostsItem extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      content: null,
    };
  }

  componentDidMount() {
    const excerpt = _get(this.props, 'node.excerpt');

    if (excerpt) {
      renderMd(excerpt)
        .then(content => this.setState({content}));
    }
  }

  render() {
    const {node} = this.props;
    if (!node) {
      return null;
    }
    const title = _get(node, 'frontmatter.title') || node.fields.slug;
    return (
      <div className='post-item'>
        <h2>
          <Link to={utils.prefixBlog(node.fields.slug)}>
            {title}
          </Link>
        </h2>
        <PostDate>
          {node.frontmatter.date}
        </PostDate>
        <p dangerouslySetInnerHTML={{__html: this.state.content}}/>
      </div>
    );
  }
}

PostsItem.propTypes = {
  node: PropTypes.shape({}),
};

PostsItem.defaultProps = {
  node: null,
};

export default PostsItem;
