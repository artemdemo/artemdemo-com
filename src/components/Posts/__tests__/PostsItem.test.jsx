import React from 'react';
import renderer from 'react-test-renderer';
import PostsItem from '../PostsItem';

jest.mock('gatsby');
jest.mock('../../../services/md');
jest.mock('../../PostDate/PostDate');

describe('PostsItem', () => {
    it('should render empty', () => {
        const tree = renderer.create(
            <PostsItem />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render with node', () => {
        const tree = renderer.create(
            <PostsItem
                node={{
                    frontmatter: {
                        title: 'Some title',
                        date: '10 May, 2018',
                    },
                    fields: {
                        slug: '/first',
                    },
                    excerpt: 'Excerpt text',
                }}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should render with slug when title not provided', () => {
        const tree = renderer.create(
            <PostsItem
                node={{
                    frontmatter: {
                        date: '10 May, 2018',
                    },
                    fields: {
                        slug: '/first',
                    },
                    excerpt: 'Excerpt text',
                }}
            />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
