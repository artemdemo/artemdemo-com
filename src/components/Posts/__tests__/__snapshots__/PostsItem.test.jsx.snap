// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsItem should render empty 1`] = `null`;

exports[`PostsItem should render with node 1`] = `
<div
  className="PostsItem__PostItemSty-sc-1fceitp-0 jEkdwx"
>
  <h2>
    <div
      data-mock="Link"
    >
      {
  "to": "/blog/first"
}
      Some title
    </div>
  </h2>
  <div
    data-mock="PostDate"
  >
    10 May, 2018
  </div>
  <p
    dangerouslySetInnerHTML={
      Object {
        "__html": "Excerpt text",
      }
    }
  />
</div>
`;

exports[`PostsItem should render with slug when title not provided 1`] = `
<div
  className="PostsItem__PostItemSty-sc-1fceitp-0 jEkdwx"
>
  <h2>
    <div
      data-mock="Link"
    >
      {
  "to": "/blog/first"
}
      /first
    </div>
  </h2>
  <div
    data-mock="PostDate"
  >
    10 May, 2018
  </div>
  <p
    dangerouslySetInnerHTML={
      Object {
        "__html": "Excerpt text",
      }
    }
  />
</div>
`;
